base/special/exp.jl

```diff
@@ -0,0 +1,141 @@
+#                       2*r
+#       exp(r) = 1 + ----------
+#                     R(r) - r
+#                          r*c(r)
+#              = 1 + r + ----------- (for better accuracy)
+#                         2 - c(r)
+#    where
+#       c(r) = r - (P1*r^2  + P2*r^4  + ... + P5*r^10 + ...).
+#
+# 3. Scale back: exp(x) = 2^k * exp(r)
+
+# log(2)
+const LN2 = 6.931471805599453094172321214581765680755001343602552541206800094933936219696955e-01
+# log2(e)
+const LOG2E = 1.442695040888963407359924681001892137426646
+
+# log(2) into upper and lower
+LN2U{T<:Float64}(::Type{T}) = T(6.93147180369123816490e-1)
+LN2L{T<:Float64}(::Type{T}) = T(1.90821492927058770002e-10)
+
+LN2U{T<:Float32}(::Type{T}) = T(6.9313812256e-1)
+LN2L{T<:Float32}(::Type{T}) = T(9.0580006145e-6)
+
+# max and min arguments for exponential fucntions
+MAXEXP(::Type{Float64}) = 7.09782712893383996732e2 # log 2^1023*(2-2^-52)
+MAXEXP(::Type{Float32}) = 88.72283905206835f0      # log 2^127 *(2-2^-23)
+
+# one less than the min exponent since we can sqeeze a bit more from the exp function
+MINEXP(::Type{Float64}) = -7.451332191019412076235e2 # log 2^-1075
+MINEXP(::Type{Float32}) = -103.97207708f0            # log 2^-150
+
+# coefficients from: lib/msun/src/e_exp.c
+@inline exp_kernel{T<:Float64}(x::T) = @horner_oftype(x, 1.66666666666666019037e-1,
+        -2.77777777770155933842e-3,
+        6.61375632143793436117e-5,
+        -1.65339022054652515390e-6,
+        4.13813679705723846039e-8)
+
+# coefficients from: lib/msun/src/e_expf.c
+@inline exp_kernel{T<:Float32}(x::T) = @horner_oftype(x, 1.6666625440e-1, -2.7667332906e-3)
+
+# custom coefficients (slightly better mean accuracy, but also a bit slower)
+# @inline exp_kernel{T<:SmallFloat}(x::T) = @horner_oftype(x, 0.1666666567325592041015625,
+#         -2.777527086436748504638671875e-3,
+#         6.451140507124364376068115234375e-5)
+
+# for values smaller than this threshold just use Taylor expansion of 1 + x
+exp_small_thres(::Type{Float64}) = 2.0^-28
+exp_small_thres(::Type{Float32}) = 2.0f0^-13
+
+"""
+    exp(x)
+
+Compute the natural base exponential of `x`, in other words ``e^x``.
+"""
+function exp{T<:Union{Float32,Float64}}(x::T)
+    xu = reinterpret(Unsigned, x)
+    xs = xu & ~sign_mask(T)
+    xsb = Int(xu >> Unsigned(sizeof(T)*8 - 1))
+
+    # filter out non-finite arguments
+    if xs > reinterpret(Unsigned, MAXEXP(T))
+        if xs >= exponent_mask(T)
+            xs & significand_mask(T) != 0 && return T(NaN)
+            return xsb == 0 ? T(Inf) : T(0.0) # exp(+-Inf)
+        end
+        x > MAXEXP(T) && return T(Inf)
+        x < MINEXP(T) && return T(0.0)
+    end
+
+    # this implementation gives 2.7182818284590455 for exp(1.0), when T ==
+    # Float64, which is well within the allowable error. however,
+    # 2.718281828459045 is closer to the true value so we prefer that answer,
+    # given that 1.0 is such an important argument value.
+    if x == T(1.0) && T == Float64
```

Comment:
Maybe `x == T(1.0) && return T(2.718281828459045235360)` so that we get the exactly rounded result for `Float32` too?